solutions/usaco-696.mdx

```diff
@@ -2,13 +2,115 @@
 id: usaco-696
 source: USACO Platinum 2017 January
 title: Promotion Counting
-author: Benjamin Qi
+author: Benjamin Qi, Timothy Gao
 ---
 
-## Merging Indexed Sets
+[Official Editorial (Java)](http://www.usaco.org/current/data/sol_promote_platinum_jan17.html)
+
+## Solution 1: Euler Tour
+
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+Recall from the [Euler tour module](/gold/tree-euler) that each subtree is represented as a contiguous interval in its Euler tour. With this knowledge, we can compress the tree into an array. Now, the problem becomes for each node $i$, look for how many elements are less than $\texttt{tour}[i]$ from $i$ to $\texttt{tout}[i]$, where $\texttt{tout}[i]$ is the end of the subtree interval for node $i$. By processing the nodes in decreasing order, this is a problem we can solve in $\mathcal{O}(N \log N)$ time with a binary-indexed tree.
+<LanguageSection>
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+public class promote {
+	
+	public static void main(String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new FileReader("promote.in"));
+		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("promote.out")));
+	 
+		int N = Integer.parseInt(br.readLine());
+		int[][] ratings = new int[N][2];
+		for (int i = 0; i < N; i++) {
+			ratings[i] = new int[]{i, Integer.parseInt(br.readLine())};
+		}
+		List<Integer>[] adj = new List[N];
+		
+		for (int i = 0; i < adj.length; i++) {
+			adj[i] = new ArrayList<>();
+		}
+		
+		for (int i = 1; i < N; i++) {
+			int par = Integer.parseInt(br.readLine()) - 1;
+			adj[par].add(i);
+		}
+		int[] ret = new int[N];
+		int[] tout = new int[N];
+		int[] bac = new int[N];
+		dfs(adj, tout, bac, 0, 0, -1);
+		BIT bit = new BIT(N);
+		Arrays.sort(ratings, (int[] a, int[] b) -> b[1] - a[1]);
+		
+		for (int i = 0; i < N;) {
+			int j = i;
+			while (j < N && ratings[i][1] == ratings[j][1]) {
+				int c = ratings[i][0];
+				ret[c] = bit.sum(tout[c]) - bit.sum(bac[c] - 1);
+				j++;
+			}
+			for (int k = i; k < j; k++) {
+				bit.update(tout[ratings[k][0]], 1);
+			}
+			i = j;
+		}
+		for(int i : ret) pw.println(i);
```

Comment:
> spaces after control statements, do this for the rest
?

Reply:
what are control statement? Made a change but not sure if it's what you're looking for.

Reply:
sorry for the late reply, that means you should add a space between control keywords and parentheses: 

correct:
```
for (int i : ret)
```

incorrect:
```
for(int i : ret)
```

this is the same for while loops, etc.

Reply:
> what are control statement? Made a change but not sure if it's what you're looking for.

yeah that change wasn't really necessary, one liners are fine

Reply:
Personally feel like an extra space is really insignificant, but made the change anyways